### ISSUE TITLE
 Add support for LFU eviction using Approximate Counting

### ISSUE THREAD

**Issue by arpitbbhayani** [DiceDB/dice#10]

**Comment by souravendra** [comment]
@arpitbbhayani can you please explain this a bit more? I'd like to pick this up as it sounds very interesting and I could possibly learn a lot from it.

**Comment by Rits1272** [comment]
@arpitbbhayani would like to take this up.

**Comment by Rits1272** [comment]
@JyotinderSingh can I pick this up?

**Comment by JyotinderSingh** [comment]
> @JyotinderSingh can I pick this up?

This is a relatively advanced feature, please go through the discussion which happened on the associated PR about the memory requirements.

Once done, let me know on this issue if you would like to try this.

**Comment by Rits1272** [comment]
@JyotinderSingh this seems interesting. Have gone through the discussion that happened in the previous PR. 
The gist is to implement LFU with highly optimal space complexity.

I am going through how other popular in-memory data stores like Redis and Caffeine (tinyLFU) have optimized their implementation on this and will come up with a draft PR for Dice after some research

**Comment by JyotinderSingh** [comment]
> @JyotinderSingh this seems interesting. Have gone through the discussion that happened in the previous PR. 
> 
> The gist is to implement LFU with highly optimal space complexity.
> 
> 
> 
> I am going through how other popular in-memory data stores like Redis and Caffeine (tinyLFU) have optimized their implementation on this and will come up with a draft PR for Dice after some research

Awesome, assigning to you.

**Comment by Rits1272** [comment]
@JyotinderSingh, here's approach I am considering:

LFU Cache Mechanism: We'll implement an LFU cache similar to Redis, leveraging the existing `LastAccessedAt` field in our codebase [here](https://github.com/DiceDB/dice/blob/efd726605d0ee2b8723c3ceea4f185c14ce79fb5/core/object.go#L10). This field will be augmented with a counter that tracks the frequency of key accesses.

Logarithmic Counter: The counter would be based on log counter for approximate counting measure, inspired by Redis's [implementation](https://github.com/redis/redis/blob/806459f481cdd0c43aeb9ba37f16154414cf5ca9/src/evict.c#L281). This will help manage the counter size efficiently.

Counter Allocation: The counter will occupy the first 8 bits of the lastAccessedAt field, while the remaining 24 bits will continue to store the UNIX timestamp, as currently done [here](https://github.com/DiceDB/dice/blob/efd726605d0ee2b8723c3ceea4f185c14ce79fb5/core/eviction.go#L40). This repurposes the first 8 bits, which are currently unused.

Decay Mechanism: We can also introduce decay mechanism to naturally age keys over time to improve efficiency

This approach does not increase space complexity and avoids the need for additional data structures. Let me know your thoughts on this

If everything good, will start working on its implementation

**Comment by Rits1272** [comment]
Gentle nudge ^ @JyotinderSingh

**Comment by JyotinderSingh** [comment]
> @JyotinderSingh, here's approach I am considering:
> 
> 
> 
> LFU Cache Mechanism: We'll implement an LFU cache similar to Redis, leveraging the existing `LastAccessedAt` field in our codebase [here](https://github.com/DiceDB/dice/blob/efd726605d0ee2b8723c3ceea4f185c14ce79fb5/core/object.go#L10). This field will be augmented with a counter that tracks the frequency of key accesses.
> 
> 
> 
> Logarithmic Counter: The counter would be based on log counter for approximate counting measure, inspired by Redis's [implementation](https://github.com/redis/redis/blob/806459f481cdd0c43aeb9ba37f16154414cf5ca9/src/evict.c#L281). This will help manage the counter size efficiently.
> 
> 
> 
> Counter Allocation: The counter will occupy the first 8 bits of the lastAccessedAt field, while the remaining 24 bits will continue to store the UNIX timestamp, as currently done [here](https://github.com/DiceDB/dice/blob/efd726605d0ee2b8723c3ceea4f185c14ce79fb5/core/eviction.go#L40). This repurposes the first 8 bits, which are currently unused.
> 
> 
> 
> Decay Mechanism: We can also introduce decay mechanism to naturally age keys over time to improve efficiency
> 
> 
> 
> This approach does not increase space complexity and avoids the need for additional data structures. Let me know your thoughts on this
> 
> 
> 
> If everything good, will start working on its implementation

Sounds great! Work on decay mechanism can be done in a separate PR. Let's first focus on gettng a stable and solid LFU implementation going.
Your plan sounds good, please proceed with it. Feel free to raise a draft PR while you work on it.

**Comment by arpitbbhayani** [comment]
Hello @Rits1272,

There has been no activity on this issue for the past 5 days.
It would be awesome if you keep posting updates to this issue so that we know you are actively working on it.

We are really eager to close this issue at the earliest, hence if we continue to see the inactivity, we will have to reassign the issue to someone else. We are doing this to ensure that the project maintains its momentum and others are not blocked on this work.

Just drop a comment with the current status of the work or share any issues you are facing. We can always chip in to help you out.

Thanks again.
